wallet/src/actors/crypto/mod.rs

```diff
@@ -1,61 +1,95 @@
+    ) -> Addr<Self> {
+        SyncArbiter::start(1, move || Self {
+            seed_password: seed_password.clone(),
+            master_key_salt: master_key_salt.clone(),
+            id_hash_iterations,
+            id_hash_function: id_hash_function.clone(),
+            rng: RefCell::new(rand::thread_rng()),
+        })
     }
 
     /// Generates the HD Master ExtendedKey for a wallet
     pub fn gen_master_key(
         &self,
-        seed_source: wallet::SeedSource,
-    ) -> Result<wallet::MasterKey, Error> {
-        let key = match seed_source.source {
-            wallet::SeedFrom::Mnemonics => {
-                let mnemonic =
-                    Mnemonic::from_phrase(seed_source.data).map_err(error::Error::WrongMnemonic)?;
-                let seed = mnemonic.seed(&self.params.seed_password);
+        seed_source: app::SeedSource,
+    ) -> Result<wallet::MasterKey, crypto::Error> {
+        let key = match seed_source {
+            app::SeedSource::Mnemonics(mnemonic) => {
+                let seed = mnemonic.seed(&self.seed_password);
 
                 MasterKeyGen::new(seed)
-                    .with_key(self.params.master_key_salt.as_ref())
+                    .with_key(self.master_key_salt.as_ref())
                     .generate()
-                    .map_err(Error::KeyGenFailed)?
+                    .map_err(crypto::Error::KeyGenFailed)?
+            }
+            app::SeedSource::Xprv => {
+                // TODO: Implement key generation from xprv
+                unimplemented!("xprv not implemented yet")
             }
-            wallet::SeedFrom::Xprv => unimplemented!(),
         };
 
         Ok(key)
     }
 
     /// Generate an ID for a wallet
     pub fn gen_id(&self, key: &wallet::MasterKey) -> String {
-        match self.params.id_hash_function {
-            wallet::HashFunction::Sha256 => {
+        match self.id_hash_function {
+            HashFunction::Sha256 => {
                 let password = [key.secret(), key.chain_code].concat();
                 let id_bytes = pbkdf2_sha256(
                     password.as_ref(),
-                    self.params.master_key_salt.as_ref(),
-                    self.params.id_hash_iterations,
+                    self.master_key_salt.as_ref(),
+                    self.id_hash_iterations,
+                );
+
+                hex::encode(id_bytes)
+            }
+        }
+    }
+
+    /// Generate a Session ID for an unlocked wallet
+    pub fn gen_session_id(&self, key: &wallet::Key) -> String {
+        match self.id_hash_function {
+            HashFunction::Sha256 => {
+                let rand_bytes: [u8; 32] = self.rng.borrow_mut().gen();
+                let password =
+                    [key.secret.as_ref(), key.salt.as_ref(), rand_bytes.as_ref()].concat();
+                let id_bytes = pbkdf2_sha256(
+                    password.as_ref(),
+                    self.master_key_salt.as_ref(),
+                    self.id_hash_iterations,
```

Comment:
@mariocao @aesedepece @girazoki 📣️ Please, can you review this function? This is the function used to generate session ids for unlocked wallets. Here `key` is the Key generated from the password used to encrypt/decrypt the wallet in the storage.

Reply:
Even if it looks overkill at first sight, I like the idea of using a password-based key derivation function so the logging-in mechanism is rate limited so as to deter bruteforce attacks.

In terms of security / privacy, my intuition is that this approach is correct. As a matter of fact, with a password that is partially or totally secret, a single hash should be enough to avoid recovery of its "ingredients". Of course, in this case a bruteforce attack may be quite feasible as the length of the ingredients is known, hence the need for thousands of PBKDF rounds. Deniability is not a problem in this case either (if someone guesses the ingredients of the password, being able to map that to a session ID does no additional harm).

@girazoki @mariocao have your say!